src/lib/root_prover/root_prover.ml

```diff
@@ -0,0 +1,160 @@
+open Protocols
+open Core
+open Async
+open Protocols.Coda_transition_frontier
+open Coda_base
+
+module type Inputs_intf = sig
+  include Transition_frontier.Inputs_intf
+
+  module Transition_frontier :
+    Transition_frontier_intf
+    with type state_hash := State_hash.t
+     and type external_transition_verified := External_transition.Verified.t
+     and type ledger_database := Ledger.Db.t
+     and type staged_ledger := Staged_ledger.t
+     and type masked_ledger := Ledger.Mask.Attached.t
+     and type transaction_snark_scan_state := Staged_ledger.Scan_state.t
+     and type staged_ledger_diff := Staged_ledger_diff.t
+     and type consensus_local_state := Consensus.Local_state.t
+
+  module Time : Protocols.Coda_pow.Time_intf
+
+  module Protocol_state_validator :
+    Protocol_state_validator_intf
+    with type time := Time.t
+     and type state_hash := State_hash.t
+     and type external_transition := External_transition.t
+     and type external_transition_proof_verified :=
+                External_transition.Proof_verified.t
+     and type external_transition_verified := External_transition.Verified.t
+end
+
+module Make (Inputs : Inputs_intf) :
+  Coda_transition_frontier.Root_prover_intf
+  with type state_body_hash := State_body_hash.t
+   and type transition_frontier := Inputs.Transition_frontier.t
+   and type external_transition := Inputs.External_transition.t
+   and type proof_verified_external_transition :=
+              Inputs.External_transition.Proof_verified.t
+   and type consensus_state := Consensus.Consensus_state.value = struct
+  open Inputs
+
+  type t = {logger: Logger.t; finality_length: int}
+
+  let create ~logger ~finality_length = {logger; finality_length}
+
+  let hash_transition =
+    Fn.compose Consensus.Protocol_state.hash External_transition.protocol_state
+
+  let get_state_hash transition_with_hash =
+    let open External_transition in
+    transition_with_hash |> With_hash.data |> Verified.protocol_state
+    |> Protocol_state.body |> Protocol_state.Body.hash
+
+  let consensus_state transition =
+    External_transition.(
+      protocol_state transition |> Protocol_state.consensus_state)
+
+  let prove ~frontier {logger; _} seen_consensus_state :
```

Comment:
This function doesn't appear to handle the case when `best_tip.length - root.length < k`. This should handle this case by accessing some historical information regarding recent previous `root` values, or by updating the overall logic to not depend on the length of the proof being equal to `k`.

Reply:
If we have `best_tip.length - root.length < k`, then the proof should return `None`. This state implies you cannot return a valid proof, so you return `None`